cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1259,6 +1259,90 @@ struct start_offset_output_iterator {
   }
 };
 
+struct flat_column_num_rows {
+  gpu::PageInfo const* pages;
+  gpu::ColumnChunkDesc const* chunks;
+
+  __device__ size_type operator()(size_type pindex)
+  {
+    gpu::PageInfo const& page = pages[pindex];
+    // ignore dictionary pages and pages belonging to any column containing repetition (lists)
+    if ((page.flags & gpu::PAGEINFO_FLAGS_DICTIONARY) ||
+        (chunks[page.chunk_idx].max_level[gpu::level_type::REPETITION] > 0)) {
+      return 0;
+    }
+    return page.num_rows;
+  }
+};
+
+struct row_counts_nonzero {
+  __device__ bool operator()(size_type count) { return count > 0; }
+};
+
+struct row_counts_different {
```

Comment:
I think this one is small enough to just be an embedded lambda.

Reply:
I normally would, but the compiler gives you the extended-lambda warning/error so I have to make them functors :(